llm/llama.go

```diff
@@ -218,366 +139,480 @@ func (ft llamaFileType) String() string {
-		llm.NumKeep = len(tokens)
+	if len(adapters) > 0 {
+		return nil, errors.New("ollama supports only one lora adapter, but multiple were provided")
 	}
 
-	cTokens := make([]C.llama_token, len(tokens))
-	for i := range tokens {
-		cTokens[i] = C.llama_token(tokens[i])
+	port, err := getAvailablePort()
+	if err != nil {
+		return nil, fmt.Errorf("load llm: %w", err)
 	}
 
-	// min(llm.NumCtx - 4, llm.NumKeep)
-	if llm.NumCtx-4 < llm.NumKeep {
-		llm.NumKeep = llm.NumCtx - 4
+	params := []string{
+		"--port", fmt.Sprintf("%d", port),
+		"--ctx-size", fmt.Sprintf("%d", opts.NumCtx),
+		"--gqa", fmt.Sprintf("%d", opts.NumGQA),
+		"--rope-freq-base", fmt.Sprintf("%f", opts.RopeFrequencyBase),
+		"--rope-freq-scale", fmt.Sprintf("%f", opts.RopeFrequencyScale),
+		"--batch-size", fmt.Sprintf("%d", opts.NumBatch),
+		"--embedding",
 	}
 
-	if len(tokens) >= llm.NumCtx {
-		// truncate input
-		numLeft := (llm.NumCtx - llm.NumKeep) / 2
-		truncated := cTokens[:llm.NumKeep]
-		erasedBlocks := (len(cTokens) - llm.NumKeep - numLeft - 1) / numLeft
-		truncated = append(truncated, cTokens[llm.NumKeep+erasedBlocks*numLeft:]...)
-		copy(llm.last, cTokens[len(cTokens)-llm.NumCtx:])
+	if opts.NumGPU > 0 {
+		params = append(params, "--n-gpu-layers", fmt.Sprintf("%d", opts.NumGPU))
+	}
 
-		cTokens = truncated
-		log.Printf("input truncated: num_ctx=%d num_keep=%d num_left=%d num_tokens=%d", llm.NumCtx, llm.NumKeep, numLeft, len(truncated))
+	if len(adapters) > 0 {
+		// TODO: applying multiple adapters is not supported by the llama.cpp server yet
+		params = append(params, "--lora", adapters[0])
+		params = append(params, "--lora-base", model)
 	} else {
-		llm.last = make([]C.llama_token, llm.NumCtx-len(cTokens))
-		llm.last = append(llm.last, cTokens...)
+		params = append(params, "--model", model)
 	}
 
-	var i int
-	for i = 0; i < len(llm.embd) && i < len(cTokens) && llm.embd[i] == cTokens[i]; i++ {
-		// noop
+	if opts.NumThread > 0 {
+		params = append(params, "--threads", fmt.Sprintf("%d", opts.NumThread))
 	}
 
-	llm.embd = cTokens
-	if i == len(cTokens) {
-		// evaluate at least one token to generate logits
-		i--
+	if !opts.F16KV {
+		params = append(params, "--memory-f32")
+	}
+	if opts.UseMLock {
+		params = append(params, "--mlock")
+	}
+	if !opts.UseMMap {
+		params = append(params, "--no-mmap")
+	}
+	if opts.UseNUMA {
+		params = append(params, "--numa")
 	}
 
-	llm.cursor = i
+	// try to start llama.cpp with gpu acceleration, if it fails, fallback to CPU
+	cmd, err := llamaCmd(llamaGPU, params)
+	if err != nil {
+		return nil, fmt.Errorf("llama.cpp gpu command setup: %w", err)
+	}
+	err = cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("error starting the external server: %w", err)
+	}
 
-	log.Printf("prompt: num_past=%d cached=%v eval=%v", i, len(llm.embd[:i]), len(llm.embd[i:]))
-	return cTokens
-}
+	// monitor the command in a goroutine, if it fails log the error and start the CPU version
+	go func() {
+		err := cmd.Wait()
```

Comment:
This shouldn't call wait. Instead, use either the PID or the port to detect if the process is running

Reply:
My goal here was to capture any possible runtime errors from the server, the port check is after this

Reply:
This can't be done because the server is a long running process